va/caa.go

```diff
@@ -40,16 +55,236 @@ func (va *ValidationAuthorityImpl) IsCAAValid(ctx context.Context, req *vapb.IsC
+		"result":       checkResult,
+		"problem_type": probType,
+	}).Observe(checkLatency.Seconds())
+
+	va.log.AuditObject("CAA check result", logEvent)
+
+	if prob != nil {
+		return &vapb.IsCAAValidResponse{Problem: &corepb.ProblemDetails{
+			ProblemType: probType,
+			Detail:      detail,
+		}}, nil
+	} else {
+		return &vapb.IsCAAValidResponse{}, nil
+	}
+}
+
+// processRemoteCAAResults evaluates a primary VA result, and a channel of
+// remote VA problems to produce a single overall validation result based on
+// configured feature flags. The overall result is calculated based on the VA's
+// configured `maxRemoteFailures` value.
+//
+// If the `MultiCAAFullResults` feature is enabled then
+// `processRemoteCAAResults` will expect to read a result from the
+// `remoteResultsChan` channel for each VA and will not produce an overall
+// result until all remote VAs have responded. In this case
+// `logRemoteDifferentials` will also be called to describe the differential
+// between the primary and all of the remote VAs.
+//
+// If the `MultiCAAFullResults` feature flag is not enabled then
+// `processRemoteCAAResults` will potentially return before all remote VAs have
+// had a chance to respond. This happens if the success or failure threshold is
+// met. This doesn't allow for logging the differential between the primary and
+// remote VAs but is more performant.
+func (va *ValidationAuthorityImpl) processRemoteCAAResults(
+	domain string,
+	acctID int64,
+	challengeType string,
+	primaryResult *probs.ProblemDetails,
+	remoteResultsChan chan *remoteVAResult) *probs.ProblemDetails {
+
+	state := "failure"
+	start := va.clk.Now()
+
+	defer func() {
+		va.metrics.remoteCAACheckTime.With(prometheus.Labels{
+			"type":   challengeType,
+			"result": state,
+		}).Observe(va.clk.Since(start).Seconds())
+	}()
+
+	required := len(va.remoteVAs) - va.maxRemoteFailures
+	good := 0
+	bad := 0
+
+	var remoteResults []*remoteVAResult
+	var firstProb *probs.ProblemDetails
+	// Due to channel behavior this could block indefinitely and we rely on gRPC
+	// honoring the context deadline used in client calls to prevent that from
+	// happening.
+	for result := range remoteResultsChan {
+		// Add the result to the slice
+		remoteResults = append(remoteResults, result)
+		if result.Problem == nil {
+			good++
+		} else {
+			bad++
+		}
+		// Store the first non-nil problem to return later (if `MultiCAAFullResults`
+		// is enabled).
+		if firstProb == nil && result.Problem != nil {
+			firstProb = result.Problem
+		}
+		// If MultiCAAFullResults isn't enabled then return early whenever the
+		// success or failure threshold is met.
+		if !features.Get().MultiCAAFullResults {
+			if good >= required {
+				state = "success"
+				return nil
+			} else if bad > va.maxRemoteFailures {
+				modifiedProblem := *result.Problem
+				modifiedProblem.Detail = "During secondary CAA checking: " + firstProb.Detail
+				return &modifiedProblem
+			}
+		}
```

Comment:
I've incorporated the `firstProb` change, but am leaving the break=>return out. The returns are necessary because the VA should make an early decision to bail if an RVA returned an error `MultiCAAFullResults = False`. A differential should not be produced in that case. If `MultiCAAFullResults = True`, the VA does wait for all RVAs to return before it produces a differential log and decision. Though, what's really the difference between configuring 1 RVA with/without `MultiCAAFullResults`?